### ISSUE TITLE
Overwriting ReadGroup ID

### ISSUE THREAD

**Issue by linhbngo** [biointec/halvade#7]
I am running the options of merging multiple bam files into a single bam file, and it works so far. However, each of the input bam files has its own ReadGroup ID that seems to be overwritten by the default GROUP1 ID. Is there a way to overcome this issue?

Also, does reading from multiple bam files and merge them into a single bam also include sorting?

Thanks, 

Linh

**Comment by ddcap** [comment]
If it reads from BAM input file it should use the readgroup id that is provided unless you used the '--update_rg' option, which will use the readgroup information provided in the options.

**Comment by linhbngo** [comment]
I apologize for not following up on this issue!

I have some more clarification on this issue. 

That is, the individual input bam files each has its own RG:
ID:X1 and ID:X2. This is checked by using samtools view -H <bam file> | grep @RG.
The merged bam file's readgroup viewed by samtools view -H is the readgroup of the first bam file, X1. This is consistent with the way the code works (BamMergeReducer.java: 54-55, HalvadeConf.java: 506-512, MapReduceRunner.java: 285-296). 

However, individual reads of the merged bam file neither carry the new assumed readgroup X1 nor original individual readgroup of X1 or X2 but the default readgroup GROUP1. My guess is that this is because in BamMergeReducer.java: 75-85, each new SAM record has its RG set to RGID, which is defaulted to the protected string ID RGID=GROUP1. 

    protected void reduce(ChromosomeRegion key, Iterable<SAMRecordWritable> values, Context context) throws IOException, InterruptedException {
        Iterator<SAMRecordWritable> it = values.iterator();
        SAMRecord sam = null;
        while(it.hasNext()) {
            sam = it.next().get();
            sam.setAttribute(SAMTag.RG.name(), RGID);
            samWritable.set(sam);
            recordWriter.write(outKey, samWritable);
                    
        }
    }

I will comment our the sam.setAttribute line and report back the result.

**Comment by linhbngo** [comment]
Yes, that was it. Commented out that line allows the resulting combined bam to retain the individual RG within each sam record.

**Comment by ddcap** [comment]
Hi,

Thanks for telling me, I looked into the code. And I didn't realise I basically implemented a bam merge tool, and I didn't take into account that BAM input can be combind with BAM output. In any case I put that set Attribute line in an if construction that now checks if the input is also BAM that it doesn't set the ReadGroup attribute. I have updated the code and will release a new version.

Thanks,
Dries
